idx/memory/find_cache.go

```diff
@@ -123,50 +155,140 @@ func (c *FindCache) PurgeAll() {
 // disable it for `backoffTime`. Future InvalidateFor calls made during
 // the backoff time will then return immediately.
 func (c *FindCache) InvalidateFor(orgId uint32, path string) {
-	c.RLock()
-	ch := c.newSeries[orgId]
+	c.Lock()
+	findCacheInvalidationsReceived.Inc()
+	defer c.Unlock()
+	if time.Now().Before(c.backoff) {
+		findCacheInvalidationsDropped.Inc()
+		return
+	}
+
 	cache, ok := c.cache[orgId]
-	c.RUnlock()
 	if !ok || cache.Len() < 1 {
+		findCacheInvalidationsDropped.Inc()
 		return
 	}
 
+	req := invalidateRequest{
+		orgId: orgId,
+		path:  path,
+	}
 	select {
-	case ch <- struct{}{}:
+	case c.invalidateReqs <- req:
 	default:
-		c.Lock()
-		c.backoff[orgId] = time.Now().Add(c.backoffTime)
-		delete(c.cache, orgId)
-		c.Unlock()
-		for i := 0; i < len(ch); i++ {
-			select {
-			case <-ch:
-			default:
-			}
-		}
-		log.Infof("memory-idx: findCache invalidate-queue full. Disabling cache for %s. num-cached-entries=%d", c.backoffTime.String(), cache.Len())
+		c.triggerBackoff()
 		return
 	}
+}
+
+// caller must hold lock!
+func (c *FindCache) triggerBackoff() {
+	log.Infof("memory-idx: findCache invalidate-queue full. Disabling cache for %s", c.backoffTime.String())
+	c.backoff = time.Now().Add(c.backoffTime)
+	findCacheBackoff.SetTrue()
+	time.AfterFunc(c.backoffTime, findCacheBackoff.SetFalse)
+	c.cache = make(map[uint32]*lru.Cache)
+	// drain queue
+	for len(c.invalidateReqs) > 0 {
```

Comment:
This can lead to a deadlock.  If processInvalidateQueue() reads the last item from the  c.invalidateReqs chan after we get the length, then we could block forever trying to read from the chan.  And because InvalidateFor is still holding a write lock no new items will be written to the chan.

Reply:
oops. fixed